src/coreclr/vm/codeman.cpp

```diff
@@ -3917,126 +3984,380 @@ StubCodeBlockKind EEJitManager::GetStubCodeBlockKind(RangeSection * pRangeSectio
+            }
+
+            // We skipped the remainder of the DWORD,
+            // so we must set startPos to the highest position of
+            // previous DWORD, unless we are already on the first DWORD
+
+            if (startPos < NIBBLES_PER_DWORD)
+                return 0;
+
+            startPos = ((startPos >> LOG2_NIBBLES_PER_DWORD) << LOG2_NIBBLES_PER_DWORD) - 1;
+
+            // Skip "headerless" DWORDS
+
+            while (pMapStart < pMap && 0 == (tmp = VolatileLoadWithoutBarrier<DWORD>(--pMap)))
+            {
+                startPos -= NIBBLES_PER_DWORD;
+            }
+
+            // This helps to catch degenerate error cases. This relies on the fact that
+            // startPos cannot ever be bigger than MAX_UINT
+            if (((INT_PTR)startPos) < 0)
+                return 0;
+
+            // Find the nibble with the header in the DWORD
+
+            while (startPos && !(tmp & NIBBLE_MASK))
+            {
+                tmp = tmp >> NIBBLE_SIZE;
+                startPos--;
+            }
+
+            if (startPos == 0 && tmp == 0)
+                return 0;
 
-    // Is there a header in the remainder of the DWORD ?
-    tmp = tmp >> NIBBLE_SIZE;
+            return base + POSOFF2ADDR(startPos, tmp & NIBBLE_MASK);
+        }
+    };
 
-    if (tmp)
+    struct ConstantLookupNibbleMap
     {
-        startPos--;
-        while (!(tmp & NIBBLE_MASK))
+        public:
+
+        // Write Algo:
+        //     1. Write nibble as in previous algorithm
+        //     2. If function completely covers following DWORDs, insert relative pointers
+        static void SetUnlocked(HeapList *pHp, TADDR pCode, size_t codeSize){
+            CONTRACTL {
+                NOTHROW;
+                GC_NOTRIGGER;
+            } CONTRACTL_END;
+
+            _ASSERTE(pCode >= pHp->mapBase);
+
+            // TODO: Is this check correct?
+            _ASSERTE(pCode + codeSize <= pHp->startAddress + pHp->maxCodeHeapSize);
+
+            // remove bottom two bits to ensure alignment math
+            // on ARM32 Thumb, the low bits indicate the thumb instruction set
+            pCode = ALIGN_DOWN(pCode, CODE_ALIGN);
+
+            size_t delta = pCode - pHp->mapBase;
+
+            size_t dwordIndex = GetDwordIndex(delta);
+            size_t nibbleIndex = GetNibbleIndex(delta);
+            DWORD value = ADDR2OFFS(delta);
+
+            DWORD mask  = POS2MASK(nibbleIndex);
+
+            value = value << Pos2ShiftCount(nibbleIndex);
+
+            PTR_DWORD pMap = pHp->pHdrMap;
+
+            // assert that we don't overwrite an existing offset
+            // the nibble is empty and the DWORD is not a pointer
+            _ASSERTE(!((*(pMap+dwordIndex)) & ~mask) && !IsPointer(*(pMap+dwordIndex)));
+
+            // It is important for this update to be atomic. Synchronization would be required with FindMethodCode otherwise.
+            *(pMap+dwordIndex) = ((*(pMap+dwordIndex)) & mask) | value;
```

Comment:
Should this be a volatile load followed by a compare and swap?

Reply:
The comment and the method of writing to a DWORD is unchanged from the original nibblemap algorithm.
I left this comment in because I wasn't 100% why it is there.

From my perspective I don't think we need explicit synchronization on the writes/reads for a couple of reasons.

1. Writes/deletes are locked behind holding the critical section so writes won't interfere with each other.
2. While it would be bad if the read algorithm reads a partially written DWORD, it should not matter if the read sees the DWORD before or after a complete write. I do have some concerns if this method of writing will give us this "complete" write behavior on all platforms, however, given that we have already been using this, I think it probably does.

If these is a safer or more pragmatic way of writing here, I'm happy to change to that.


See: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/codeman.cpp#L4068-L4069